### PR TITLE
Add `gdb` pretty-printers for simple types

### PR DESCRIPTION
**Pull request by upsj**
## Description
This adds `gdb` pretty printers for `rmm::device_uvector`, `thrust::*_vector`, `thrust::device_reference` and `cudf::*_span`. The implementation is based on https://github.com/NVIDIA/thrust/pull/1631. I will probably backport the thrust-specific changes to there as well, but since the location of the thrust source is not fixed, I'd prefer having all types in a self-contained file.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by upsj** [comment]
This seems to be working for all the cases I tested. The csv reader is a good place to try them out, since it uses host_span, device_span and device_uvector.

**Comment by harrism** [comment]
Is this the right home for all of this? You already added pretty printers for thrust in thrust. Pretty printers for RMM should probably live in RMM so more of RAPIDS can benefit?

**Comment by upsj** [comment]
I believe it would be hard to import dependencies from other components, since they may be placed in an arbitrary location (build/$config/_deps/.../scripts/gdb-pretty-printers.py in rapids-compose if pulling it via CPM, somewhere else if using an existing library). If I wanted to use pretty-printers from the individual files, I'd have to run
```
source path/to/rmm/scripts/gdb-pretty-printer.py
source path/to/thrust/scripts/gdb-pretty-printer.py
source path/to/cudf/scripts/gdb-pretty-printer.py
```
manually instead of just sourcing a single file.
All pretty-printers here rely on `DeviceIterator` and `HostIterator`, so they need to either pull in the file containing them or we duplicate the code across all of them.

OTOH, you mentioned `thrust::*_vector` no longer being used inside cudf, so maybe this isn't too much of a roadblock after all?

**Comment by harrism** [comment]
Selfishly, I want to be able to find RMM pretty printers in RMM so that I can use them there and in other libraries. So I guess it's OK to duplicate them in cuDF, but I think they should be added to RMM as well.

**Comment by davidwendt** [comment]
I'm also in favor of putting these in their respective repos. Perhaps a script can be used to wrap the

> I believe it would be hard to import dependencies from other components, since they may be placed in an arbitrary location (build/$config/_deps/.../scripts/gdb-pretty-printers.py in rapids-compose if pulling it via CPM, somewhere else if using an existing library). If I wanted to use pretty-printers from the individual files, I'd have to run
> 
> ```
> source path/to/rmm/scripts/gdb-pretty-printer.py
> source path/to/thrust/scripts/gdb-pretty-printer.py
> source path/to/cudf/scripts/gdb-pretty-printer.py
> ```
> 
> manually instead of just sourcing a single file.

Could we not just have a script to do these calls?  I'd rather not have duplicates to maintain.

**Comment by upsj** [comment]
I totally see the issue - we could configure the rmm and thrust path via CMake and create the pretty-printer inside `build`. Would that be an acceptable solution?

**Comment by davidwendt** [comment]
> I totally see the issue - we could configure the rmm and thrust path via CMake and create the pretty-printer/load script inside `build`. Would that be an acceptable solution?

I think that makes sense. These are only useful if you do a debug build after all.

**Comment by upsj** [comment]
Considering the horrible things I have to do to try and make the deduplication work (concatenating files in CMake! `exec(open(...).read())` in Python, I'm very open to hearing alternative approaches. `gdb` has its own [command file syntax](https://sourceware.org/gdb/onlinedocs/gdb/Command-Files.html) that could be used to load the individual files one after the other.

**Comment by davidwendt** [comment]
> Considering the horrible things I have to do to try and make the deduplication work (concatenating files in CMake! `exec(open(...).read())` in Python), I'm very open to hearing alternative approaches. `gdb` has its own [command file syntax](https://sourceware.org/gdb/onlinedocs/gdb/Command-Files.html) that could be used to load the individual files one after the other.

The gdb approach seems reasonable. Perhaps we could create some documentation on debugging and include a gdb helper script to illustrate how to load/use these.

**Comment by upsj** [comment]
I think this is in a good state for a review now. Sadly, since `rmm` is being installed by default inside `rapids-compose`, CPM picks up the installed `rmm` and thus the `load-pretty-printers` file doesn't get created there.

**Comment by karthikeyann** [comment]
style check cmake-format.............................................................Failed

**Comment by upsj** [comment]
yes, this depends on https://github.com/rapidsai/rmm/pull/1088 being merged, so I removed the incorrect label

File: cpp/scripts/gdb-pretty-printers.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import gdb
+
+
+if "HostIterator" not in locals():
+    raise Exception("This file expects the RMM pretty-printers to be loaded already. "
```
Comment (bdice) [inline cpp/scripts/gdb-pretty-printers.py:20]:
```suggestion
    raise NameError("This file expects the RMM pretty-printers to be loaded already. "
```

File: cpp/scripts/gdb-pretty-printers.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        self.val = val
+        self.pointer = val["_data"]
+        self.size = int(val["_size"])
+
+    def children(self):
+        return HostIterator(self.pointer, self.size)
+
+    def to_string(self):
+        typename = str(self.val.type)
+        return ("{typename} of length {self.size} at {hex(self.pointer)}")
```
Comment (bdice) [inline cpp/scripts/gdb-pretty-printers.py:39]:
```suggestion
        return f"{self.val.type} of length {self.size} at {hex(self.pointer)}"
```

File: cpp/scripts/gdb-pretty-printers.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        self.val = val
+        self.pointer = val["_data"]
+        self.size = int(val["_size"])
+
+    def children(self):
+        return DeviceIterator(self.pointer, self.size)
+
+    def to_string(self):
+        typename = str(self.val.type)
+        return ("{typename} of length {self.size} at {hex(self.pointer)}")
```
Comment (bdice) [inline cpp/scripts/gdb-pretty-printers.py:58]:
```suggestion
        return f"{self.val.type} of length {self.size} at {hex(self.pointer)}"
```

File: cpp/scripts/gdb-pretty-printers.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import gdb
+
+
+if "HostIterator" not in locals():
```
Comment (bdice) [inline cpp/scripts/gdb-pretty-printers.py:19]: Make an explicit list of the names this relies on.
```suggestion
if not all(name in locals() for name in ("HostIterator", "DeviceIterator", "is_template_type_not_alias", "template_match")):
```

File: cpp/scripts/gdb-pretty-printers.py [hunk 5]
```diff
@@ -35,8 +34,7 @@ def children(self):
         return HostIterator(self.pointer, self.size)
 
     def to_string(self):
-        typename = str(self.val.type)
-        return ("{typename} of length {self.size} at {hex(self.pointer)}")
+        return ("{self.val.type} of length {self.size} at {hex(self.pointer)}")
```
Comment (bdice) [inline cpp/scripts/gdb-pretty-printers.py:37]:
```suggestion
        return f"{self.val.type} of length {self.size} at {hex(self.pointer)}"
```

File: cpp/scripts/gdb-pretty-printers.py [hunk 6]
```diff
@@ -54,8 +52,7 @@ def children(self):
         return DeviceIterator(self.pointer, self.size)
 
     def to_string(self):
-        typename = str(self.val.type)
-        return ("{typename} of length {self.size} at {hex(self.pointer)}")
+        return ("{self.val.type} of length {self.size} at {hex(self.pointer)}")
```
Comment (bdice) [inline cpp/scripts/gdb-pretty-printers.py:55]:
```suggestion
        return f"{self.val.type} of length {self.size} at {hex(self.pointer)}"
```

**Review by bdice — APPROVED** [review]
Nice work!

File: cpp/scripts/gdb-pretty-printers.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    def to_string(self):
+        return f"{self.val.type} of length {self.size} at {hex(self.pointer)}"
+
+    def display_hint(self):
+        return "array"
+
+
+def lookup_cudf_type(val):
+    if not str(val.type.unqualified()).startswith("cudf::"):
```
Comment (karthikeyann) [inline cpp/scripts/gdb-pretty-printers.py:72]: does this printer work if we use `using namespace cudf` ?

**Review by karthikeyann — APPROVED** [review]
LGTM 👍

File: cpp/scripts/gdb-pretty-printers.py (hunk 7 above)
Comment (upsj) [reply cpp/scripts/gdb-pretty-printers.py:72]: yes, `using` is a source-level alias that doesn't affect the generated debug info, afaik.

**Comment by upsj** [comment]
@gpucibot merge

File: cpp/scripts/gdb-pretty-printers.py (hunk 7 above)
Comment (upsj) [reply cpp/scripts/gdb-pretty-printers.py:72]: I can confirm that's the case after some small experiments :)

**Pull request merged**